### PR TITLE
Improve Shaders

### PR DESCRIPTION
**Pull request by tinxx**
-- Due to an oversight on my part, this MR is based on #804 – so please wait until #804 is merged. --

This MR improves the portal shaders on various aspects:

- The inside of clipped nodes is visible.
- Text shaders honor portals again
  -  The updated coordinates are used for the portal effect after billboarding is applied so that the portal effect now works as expected.
- Both points that define the portal rect are stored in one shader property.
  - Shader `Vector` properties always use four dimensions, so it is more efficient to store both 2D coords in one property.
- Files using both tabs and spaces for indentations have been changed to use either one (but not both).

### CODE REVIEW

File: Assets/Resources/Shaders/TMP_SDF-Mobile Portal Overlay.shader [hunk 1]
```diff
@@ -149,7 +147,27 @@ SubShader {
 			float4 vert = input.vertex;
 			vert.x += _VertexOffsetX;
 			vert.y += _VertexOffsetY;
+
+#ifdef SEE_TEXT_FACING_CAMERA
+			// Apply billboarding in view-space
+			// The following two variables will make sure that the localScale is correctly applied to the billboarded text.
+			// Source: https://forum.unity.com/threads/billboard-shader-that-respects-gameobjects-transform-localscale.451431
+			float scaleX = length(float4(UNITY_MATRIX_M[0].r, UNITY_MATRIX_M[1].r, UNITY_MATRIX_M[2].r, UNITY_MATRIX_M[3].r));
+            float scaleY = length(float4(UNITY_MATRIX_M[0].g, UNITY_MATRIX_M[1].g, UNITY_MATRIX_M[2].g, UNITY_MATRIX_M[3].g));
```
Comment (falko17) [inline Assets/Resources/Shaders/TMP_SDF-Mobile Portal Overlay.shader:156]: There's still some mixed tabs/spaces here.

File: Assets/Resources/Shaders/TMP_SDF-Mobile Portal.shader [hunk 2]
```diff
@@ -147,7 +145,27 @@ SubShader {
 			float4 vert = input.vertex;
 			vert.x += _VertexOffsetX;
 			vert.y += _VertexOffsetY;
+
+#ifdef SEE_TEXT_FACING_CAMERA
+			// Apply billboarding in view-space
+			// The following two variables will make sure that the localScale is correctly applied to the billboarded text.
+			// Source: https://forum.unity.com/threads/billboard-shader-that-respects-gameobjects-transform-localscale.451431
+			float scaleX = length(float4(UNITY_MATRIX_M[0].r, UNITY_MATRIX_M[1].r, UNITY_MATRIX_M[2].r, UNITY_MATRIX_M[3].r));
+            float scaleY = length(float4(UNITY_MATRIX_M[0].g, UNITY_MATRIX_M[1].g, UNITY_MATRIX_M[2].g, UNITY_MATRIX_M[3].g));
```
Comment (falko17) [inline Assets/Resources/Shaders/TMP_SDF-Mobile Portal.shader:154]: See above: Mixed spaces/tabs.

File: Assets/Resources/Shaders/TransparentLinePortalShader.shader [hunk 3]
```diff
@@ -35,68 +34,68 @@
...
+                    discard;
+                }
                 fixed4 c = SampleSpriteTexture (IN.texcoord) * IN.color;
-				if (IN.worldPos.x < _PortalMin.x || IN.worldPos.z < _PortalMin.y ||
-					IN.worldPos.x > _PortalMax.x || IN.worldPos.z > _PortalMax.y
-				)
-				{
-					c.a = 0.0f;
-				}
                 c.rgb *= c.a;
```
Comment (falko17) [inline Assets/Resources/Shaders/TransparentLinePortalShader.shader:99]: Do we still need this line, now that we no longer use the alpha value to hide fragments outside the portal?

File: Assets/SEE/Game/Portal.cs [hunk 4]
```diff
@@ -78,27 +73,25 @@ public static void GetPortal(GameObject gameObject, out Vector2 leftFront, out V
...
-        /// <see cref="portalMin"/> and <see cref="portalMax"/>. If they do not
-        /// exist, the result is undefined.
+        /// Returns the portal of <paramref name="material"/>, a rectangle on the x/z plane,
+        /// defined by its <paramref name="leftFront"/> and <paramref name="rightBack"/> corners.
+        /// <para>
+        /// Precondition: <paramref name="material"/> must have the <c>_Portal</c> attribute
+        /// If it does not exist, the result is undefined.
+        /// </para>
         /// </summary>
         /// <param name="gameObject">game objects whose portal is requested</param>
```
Comment (falko17) [inline Assets/SEE/Game/Portal.cs:83]: This isn't part of your changes, but the parameter documentation does not match the actual parameter.

**Review by falko17 — CHANGES_REQUESTED** [review]
Just four small comments this time. Honestly, it makes a lot of sense to use a single property for the portal (since it has four components anyway), I don't know why we didn't do it like this before.

File: Assets/Resources/Shaders/TransparentLinePortalShader.shader (hunk 3 above)
Comment (tinxx) [reply Assets/Resources/Shaders/TransparentLinePortalShader.shader:99]: I thought it was part of the alpha blending. Transparency can be applied via texture.

If you know more, or want to spend some time investigating, please feel free :)

Comment (falko17) [reply Assets/Resources/Shaders/TransparentLinePortalShader.shader:99]: > I thought it was part of the alpha blending. Transparency can be applied via texture.

That makes sense, I forgot that the texture can apply transparency too, so let's keep it as-is.

**Comment by falko17** [comment]
This PR should also be ready now, but I'll wait until #804 is merged before approval, so that we don't accidentally merge before then.

**Review by falko17 — APPROVED** [review]

**Pull request merged**